cli/src/cast.rs

```diff
@@ -710,6 +728,44 @@ where
     })
 }
 
+fn det_base_in(value: &str, base_in: Option<String>) -> eyre::Result<u32> {
+    match base_in {
+        Some(base_in) => match base_in.as_str() {
+            "10" | "dec" => Ok(10),
+            "16" | "hex" => Ok(16),
+            _ => eyre::bail!("Unknown input base: {base_in}"),
+        },
+        None if value.starts_with("0x") => Ok(16),
+        None => match U256::from_str_radix(value, 10) {
+            Ok(_) => {
+                U256::from_str_radix(value, 16)
+                    .expect("could not parse base in, please prepend with 0x if hex.");
```

Comment:
If the value can be decoded as base 10, then it can always be decoded as base 16, so this check is redundant.

However, this also does not solve the original issue we tried to solve: if it can be decoded as base 10 **and** it can be decoded as base 16, then it is ambigious, and that is the point we should error, but we just assume base 10. Does that make sense?

Also the error should be something like:

"Could not autodetect input base: input could be decimal or hexadecimal. Please prepend with 0x if the input is hexadecimal."

Reply:
IMO, we should be more explicit than this message, and add a ", or specify --base-in parameter" because the user may get confused if he is just trying to shift a decimal number.

Reply:
Also, I've seen in other cast utilities commands a panic in the case when an user input could not be parsed because the program would panic, and display this:

```
The application panicked (crashed).
Message:  Could not autodetect input base.: FromStrRadixErr { kind: InvalidCharacter, source: Some(Hex(FromHexError { inner: InvalidHexCharacter { c: 't', index: 2 } })) }
```

Should we refactor all places where the code could panic because of an `unwrap()` or `expect()` after parsing of the user input in order to just send the friendly message, or is that fine that way ?

Reply:
Yes, but that is part of #2131. Only making you do it for this PR so we have less work once we get around to it :)

Reply:
Lemme know if I can help